middleware/file/zone.go

```diff
@@ -24,6 +24,8 @@ type Zone struct {
 	*tree.Tree
 	Apex Apex
 
+	dnames []*dns.DNAME
```

Comment:
DNAME RRs are still kept in the tree. This slice serves as a cache which keeps another copy of those (pointers to) DNAMEs. If it is not empty then we have DNAMEs in the zone.

The RFC forbids the existence of RR at any subdomain of the owner of a DNAME ([Section 2.4](https://tools.ietf.org/html/rfc6672#section-2.4)). If a zone has DNAME(s), we will have to lookup the DNAMEs to make sure we do not give out those offending subdomains. This check is required for every single query, no matter we found a match in the tree or not.